pkg/ruler/compat.go

```diff
@@ -163,6 +163,33 @@ func MetricsQueryFunc(qf rules.QueryFunc, queries, failedQueries prometheus.Coun
 	}
 }
 
+func RecordAndReportRuleQueryMetrics(qf rules.QueryFunc, queryTime prometheus.Counter, logger log.Logger) rules.QueryFunc {
+	if queryTime == nil {
+		return qf
+	}
+
+	return func(ctx context.Context, qs string, t time.Time) (promql.Vector, error) {
+		// If we've been passed a counter we want to record the wall time spent executing this request.
+		timer := prometheus.NewTimer(nil)
+		defer func() {
+			querySeconds := timer.ObserveDuration().Seconds()
+			queryTime.Add(querySeconds)
+
+			// Log ruler query stats.
+			logMessage := []interface{}{
+				"msg", "query stats",
+				"component", "ruler",
```

Comment:
As already mentioned in the previous comment, please add `"component", "query-frontend"` to the query-frontend "query stats" logging too, so that it's more clear.